mono/metadata/object-internals.h

```diff
@@ -429,6 +429,16 @@ typedef enum {
 
 struct _MonoThreadInfo;
 
+typedef struct {
+		gboolean freeing : 1; // Either 0x1 or 0x0
+		gint32 waiters : 31;
+} MonoInternalThreadLockFlags;
+
+typedef union {
+	gint32 placeholder;
+	MonoInternalThreadLockFlags value;
+} MonoInternalThreadLockFlagBox;
+
```

Comment:
```
typedef struct {
+		gboolean freeing : 1; // Either 0x1 or 0x0
+		gint32 waiters : 31;
+} MonoInternalThreadLockFlags;
+
+typedef union {
+	gint32 placeholder;
+	MonoInternalThreadLockFlags value;
+} MonoInternalThreadLockFlagBox;

This should be combined into:

union {
  gint32 i;
  struct {
    gboolean freeing : 1:
    gint32 waiters : 31;
} MonoInternalThreadLockFlags;

but I question if 31 bits is convincingly enough on a 64bit system.
And the uncertainty is not worth it.
So like I had, plus expansion for same-size on all systems.

native:

union {
  gsize i;
  gint64 same_size_on_all_systems;
  struct {
    gboolean freeing : 1:
    gsize waiters :  (sizeof (gsize) * 8) - 1;
} MonoInternalThreadLockFlags;


managed:
  long thread_lock_flags;
```

Reply:
The combination makes it really unwieldy to work with the memory from the callers due to C's unwillingness to cast from an int to a struct.

Reply:
No it doesn't. I guess I have not been clear enough.
```
MonoInternalThreadLockFlags flags;
flags.freeing = true
flags.i  = cas(&thread->flags.i);
```

etc.